### ISSUE TITLE
Add rebase support

### ISSUE THREAD

**Issue by ethomson** [libgit2/libgit2#2253]
We have a new contributor, @jacquesg, who is interested in seeing rebase in the library, as am I.  I've just barely started down this road, but I wanted to open an issue to coordinate a bit without stepping on each others toes.  This is mostly brainstorming, so please feel free to edit / correct / flame me.

I think that we can break up rebase into these component pieces:
1. Setting up a rebase: laying down all the patches that will be involved in this, and the necessary metadata files.
2. Applying the next patch in a rebase into the workdir / index.
3. Committing the changes from (2).  I'm breaking this out into a separate step to allow conflict resolution, or handling things like `edit` / `reword`.
4. Cleaning up rebase data produced by (1).

It sounds so simple when you write it down like that!  :)  I'm missing a bunch of salient details, obviously.

Some salient details include:
- It would be nice to have a generic sequencer that can be used for rebase and the other sequenced commands.

@jacquesg, it sounded like you were interested in this work as well.  (Though I don't want to put words into your mouth or try to coerce you into working on this if you don't want to.  Well, at least not _too much_ coercion.)

I have started looking at patch application.  Since large chunks were taken from git.git I was going through the authors and planning on getting relicensing approval from them before pushing anything up.  For rebase, my thinking is that we don't need to worry about being able to apply any random patch that was generated by any random differ like `git-apply` does.  We need only to worry about git-style patches created by git.git and ourselves.

This leaves us with a incomplete patch application implementation, but it gets us to unblocking rebase more quickly.  My thinking is that we can come back and flesh out full implementation with all the crazy patches in the wild later.
